crates/providers/http-client/src/lib.rs

```diff
@@ -0,0 +1,155 @@
+use std::collections::HashMap;
+use std::str::FromStr;
+
+use async_trait::async_trait;
+use http::{HeaderMap, HeaderName, HeaderValue};
+use tracing::{error, instrument, trace, warn};
+use wasmcloud_provider_sdk::{
+    core::LinkDefinition,
+    error::{ProviderInvocationError, ProviderInvocationResult},
+    Context,
+};
+
+wasmcloud_provider_wit_bindgen::generate!({
+    impl_struct: HttpClientProvider,
+    contract: "wasmcloud:httpclient",
+    replace_witified_maps: true,
+    wit_bindgen_cfg: "provider-http-client"
+});
+
+/// HTTP client capability provider implementation struct
+#[derive(Default, Clone)]
+pub struct HttpClientProvider;
+
+/// Implement the [httpclient contract](https://github.com/wasmCloud/interfaces/blob/main/httpclient)
+/// represented by the WIT interface @ `wit/provider-httpclient.wit`
+#[async_trait]
+impl WasmcloudHttpclientHttpClient for HttpClientProvider {
+    /// Accepts a request from an actor and forwards it to a remote http server.
+    ///
+    /// This function returns an RpcError if there was a network-related
+    /// error sending the request. If the remote server returned an http
+    /// error (status other than 2xx), returns Ok with the status code and
+    /// body returned from the remote server.
+    #[instrument(level = "debug", skip(self, _ctx, req), fields(actor_id = ?_ctx.actor, method = %req.method, url = %req.url))]
+    async fn request(
+        &self,
+        _ctx: Context,
+        req: HttpRequest,
+    ) -> ProviderInvocationResult<HttpResponse> {
+        let headers: HeaderMap = build_http_header_map(&req.headers)?;
+
+        let method = reqwest::Method::from_str(&req.method).map_err(|e| {
+            ProviderInvocationError::Provider(format!("method: {}:{}", &req.method, e))
+        })?;
+
+        trace!("forwarding {} request to {}", &req.method, &req.url);
+        let response = reqwest::Client::new()
+            .request(method, &req.url)
+            .headers(headers)
+            .body(req.body)
+            .send()
+            .await
+            .map_err(|e| {
+                // send() can fail if there was an error while sending request,
+                // a redirect loop was detected, or redirect limit was exhausted.
+                // For now, we'll return an error (not HttpResponse with error
+                // status) and the caller should receive an error
+                // (needs to be tested).
```

Comment:
It looks like this was migrated from the existing code, but I'm trying to figure out what "needs to be tested" means. Surely we've checked this behavior? 😅

Reply:
Yeah I assume this is how the behavior was found... Maybe worth trying to add a test that does this explicitly in a follow up?

Reply:
Made [an issue](https://github.com/wasmCloud/wasmCloud/issues/995) for this